### PR TITLE
Add flags to build multiarch Python SDK containers and test

### PR DESCRIPTION
**Pull request by celeste-zeng**
- Use new flags to build and push multi-arch Python containers(will not affect the current building of x86 containers if flags are not specified during build);
- Add a validatescontainer arm script so it builds and pushes a multi-arch python container, and use it to run a dataflow wordcount pipeline;
- Change the script to adjust base image requirements for Python SDK containers on ARM.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          pull_request_url: ${{ github.event.issue.pull_request.url }}
+          github_repository: ${{ github.repository }}
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          github_job: "${{ github.job }} (${{ matrix.python_version }})"
+          github_current_run_id: ${{ github.run_id }}
+      - name: Install Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python_version }}
+      - name: Install Java
```
Comment (tvalentyn) [inline .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml:84]: Do we need this?

File: sdks/python/apache_beam/examples/wordcount_it_test.py [hunk 2]
```diff
@@ -111,6 +111,19 @@ def test_wordcount_it_with_prebuilt_sdk_container_cloud_build(self):
   @pytest.mark.it_validatescontainer
   def test_wordcount_it_with_use_sibling_sdk_workers(self):
     self._run_wordcount_it(wordcount.run, experiment='use_sibling_sdk_workers')
+  
+  @pytest.mark.it_postcommit
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/wordcount_it_test.py:115]: We need to remove pytest.mark.it_postcommit, otherwise regular postcommit suite will pick it up.

File: sdks/python/container/run_validatescontainer_arm.sh [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    --project=$PROJECT \
+    --region=$REGION \
+    --sdk_container_image=$CONTAINER:$TAG \
+    --staging_location=$GCS_LOCATION/staging-dataflow-arm-test \
+    --temp_location=$GCS_LOCATION/temp-dataflow-arm-test \
+    --output=$GCS_LOCATION/output \
+    --sdk_location=$SDK_LOCATION \
+    --num_workers=1 \
+    --docker_registry_push_url=$PREBUILD_SDK_CONTAINER_REGISTRY_PATH \
+    --machine_type=t2a-standard-1"
```
Comment (tvalentyn) [inline sdks/python/container/run_validatescontainer_arm.sh:112]: test already sets it.

File: sdks/python/container/run_validatescontainer_arm.sh [hunk 4]
```diff
@@ -0,0 +1,114 @@
+#!/bin/bash
```
Comment (tvalentyn) [inline sdks/python/container/run_validatescontainer_arm.sh:1]: is there any opportunity for code reuse, perhaps by parameterizing existing validatescontainer script, or moving some parts of code to gh actions?

Comment (celeste-zeng) [reply sdks/python/container/run_validatescontainer_arm.sh:1]: Tried to add some if else statement to the validatescontainer script so that both validatesContainerARM and validatesContainer use the same script. If those changes work, will remove this new script.

File: .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml (hunk 1 above)
Comment (celeste-zeng) [reply .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml]: Removed

File: sdks/python/apache_beam/examples/wordcount_it_test.py [hunk 5]
```diff
@@ -112,6 +112,17 @@ def test_wordcount_it_with_prebuilt_sdk_container_cloud_build(self):
   def test_wordcount_it_with_use_sibling_sdk_workers(self):
     self._run_wordcount_it(wordcount.run, experiment='use_sibling_sdk_workers')
 
+  @pytest.mark.it_dataflow_arm
+  def test_wordcount_fnapi_it_arm(self):
+    self._run_wordcount_it(wordcount.run, machine_type='t2a-standard-1')
+
+  @pytest.mark.it_dataflow_arm
+  def test_wordcount_it_with_use_sibling_sdk_workers_arm(self):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/wordcount_it_test.py:120]: we don't need this.

File: sdks/python/container/run_validatescontainer_arm.sh (hunk 4 above)
Comment (celeste-zeng) [reply sdks/python/container/run_validatescontainer_arm.sh:1]: As precommit tests pass, should I just change run_validatescontainer.sh and remove this new script?

File: sdks/python/container/run_generate_requirements.sh [hunk 6]
```diff
@@ -104,4 +104,16 @@ EOT
...
 pip freeze | grep -v pkg_resources >> "$REQUIREMENTS_FILE"
+
+# Check if the generated .txt file has the line "tensorflow==".
+if grep -q "tensorflow==" "$REQUIREMENTS_FILE"; then
+
+# Get the version of tensorflow from the .txt file.
+VERSION=$(grep -Po "tensorflow==\K[^;]+" "$REQUIREMENTS_FILE")
+
+# Write the line tensorflow-cpu-aws==${VERSION};platform_machine=="aarch64" to the .txt file.
+echo "tensorflow-cpu-aws==${VERSION};platform_machine==\"aarch64\"" >> "$REQUIREMENTS_FILE"
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/container/run_generate_requirements.sh:115]: also suggestion to keep both entries in the same place:

```
if grep -q "tensorflow==" "$REQUIREMENTS_FILE"; then
  # Get the version of tensorflow from the .txt file.
  TF_VERSION=$(grep -Po "tensorflow==\K[^;]+" "$REQUIREMENTS_FILE")
  TF_ENTRY="tensorflow==${TF_VERSION}"
  TF_AARCH64_ENTRY="tensorflow-cpu-aws==${TF_VERSION};platform_machine==\"aarch64\""
  sed -i "s/${TF_ENTRY}/${TF_ENTRY}\n${TF_AARCH64_ENTRY}/g" $REQUIREMENTS_FILE
fi
```

[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/container/run_generate_requirements.sh:115]: note: this portion can be committed in a separate PR if that's easier, and requirements regenerated.

[the remark this answers is not in the document]
Comment (celeste-zeng) [reply sdks/python/container/run_generate_requirements.sh]: Added this change and reran `./gradlew :sdks:python:container:py38:generatePythonRequirements
` to test, worked!

File: sdks/python/container/run_validatescontainer.sh [hunk 7]
```diff
@@ -70,22 +71,24 @@ docker -v
...
-    # Tag the docker container.
-    docker tag "apache/$IMAGE_NAME:$SDK_VERSION" "$CONTAINER:$TAG"
-
-    # Push the container
-    gcloud docker -- push $CONTAINER:$TAG
-  fi
+  # Push the container
+  gcloud docker -- push $CONTAINER:$TAG
+else
+  # Reset the test suite tag to run ARM pipelines.
```
Comment (tvalentyn) [inline sdks/python/container/run_validatescontainer.sh:88]: add a comment that ARM suite runs on GH actions only, so we don't push containers here.

Comment (celeste-zeng) [reply sdks/python/container/run_validatescontainer.sh:88]: Also added a comment on how to test ARM Python SDK containers.

File: .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+          export_default_credentials: true
+      - name: GCloud Docker credential helper
+        run: |
+          gcloud auth configure-docker us-central1-docker.pkg.dev
+      - name: Set PY_VER_CLEAN
+        id: set_py_ver_clean
+        run: |
+          PY_VER=${{ matrix.python_version }}
+          PY_VER_CLEAN=${PY_VER//.}
+          echo "py_ver_clean=$PY_VER_CLEAN" >> $GITHUB_OUTPUT
```
Comment (tvalentyn) [inline .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml:105]: nit: debugging leftover?

File: .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      - name: Set current time as docker container image tag
+        id: set_tag
+        run: |
+          echo "buildtime=$(date +'%Y%m%d-%H%M%S%N')" >> $GITHUB_OUTPUT
```
Comment (tvalentyn) [inline .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml:109]: nit: debugging leftover?

File: sdks/python/container/run_validatescontainer.sh [hunk 10]
```diff
@@ -24,10 +24,13 @@
...
 #
 # Execute from the root of the repository:
-#     test Python3.8 container:
+#     test Python3.8 x86 container:
 #         ./gradlew :sdks:python:test-suites:dataflow:py38:validatesContainer
-#     or test all supported python versions together:
+#     or test all supported python versions x86 containers together:
 #         ./gradlew :sdks:python:test-suites:dataflow:validatesContainer
+#
+# Note: ARM test suites only run on github actions. For example, to test Python3.8 ARM containers,
```
Comment (tvalentyn) [inline sdks/python/container/run_validatescontainer.sh:32]: I would make platform an argument to this script, or at least set the environment variable in gradle, instead of in GH actions. That would make it easier to trigger these tests locally, and would make it easier for us to pivot to test on Jenkins if we don't sort out pushing to GCR from GH actions in time.

File: sdks/python/test-suites/dataflow/common.gradle [hunk 11]
```diff
@@ -348,6 +348,21 @@ task validatesContainer() {
   }
 }
 
+task validatesContainerARM() {
+  def pyversion = "${project.ext.pythonVersion.replace('.', '')}"
+  dependsOn 'initializeForDataflowJob'
+  dependsOn ":sdks:python:container:py${pyversion}:docker"
+  def runScriptsPath = "${rootDir}/sdks/python/container/run_validatescontainer.sh"
```
Comment (tvalentyn) [inline sdks/python/test-suites/dataflow/common.gradle:355]: as mentioned in the other comment, I think it would be better to configure the script to run in  ARM mode around here.

File: sdks/python/container/run_generate_requirements.sh [hunk 12]
```diff
@@ -104,4 +104,13 @@ EOT
 # Remove pkg_resources to guard against
 # https://stackoverflow.com/questions/39577984/what-is-pkg-resources-0-0-0-in-output-of-pip-freeze-command
 pip freeze | grep -v pkg_resources >> "$REQUIREMENTS_FILE"
+
+if grep -q "tensorflow==" "$REQUIREMENTS_FILE"; then
```
Comment (tvalentyn) [inline sdks/python/container/run_generate_requirements.sh:108]: i recommend to move changes to this file into a separate PR.

File: .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml (hunk 8 above)
Comment (celeste-zeng) [reply .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml:105]: I think all python GH tests now has this step, added by https://github.com/apache/beam/pull/27712

File: .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml (hunk 9 above)
Comment (celeste-zeng) [reply .github/workflows/beam_Python_ValidatesContainer_Dataflow_ARM.yml:109]: I wasn't sure whether the previous edits will successfully set the current time as the tag, but based on what I found, this should work. But I can revert to the old edits first, if it doesn't work, I will then use this.

File: sdks/python/container/run_generate_requirements.sh (hunk 12 above)
Comment (celeste-zeng) [reply sdks/python/container/run_generate_requirements.sh]: Done. https://github.com/apache/beam/pull/27789

File: sdks/python/test-suites/dataflow/common.gradle (hunk 11 above)
Comment (celeste-zeng) [reply sdks/python/test-suites/dataflow/common.gradle:355]: Done.

File: sdks/python/container/run_validatescontainer.sh (hunk 10 above)
Comment (celeste-zeng) [reply sdks/python/container/run_validatescontainer.sh:32]: Added.

File: sdks/python/container/run_validatescontainer.sh [hunk 13]
```diff
@@ -74,7 +75,7 @@ CONTAINER=us.gcr.io/$PROJECT/$USER/$IMAGE_NAME
 PREBUILD_SDK_CONTAINER_REGISTRY_PATH=us.gcr.io/$PROJECT/$USER/prebuild_python${PY_VERSION//.}_sdk
 echo "Using container $CONTAINER"
 
-if [[ "$ARCH" != "ARM" ]]; then
+if [[ "$ARCH" == "x86" ]]; then
```
Comment (tvalentyn) [inline sdks/python/container/run_validatescontainer.sh:80]: it would be more reliable to check for ARM in line 86 and error-out if the value is unrecognized.

**Review by tvalentyn — APPROVED** [review]
